.github/workflows/benchmark.yaml

```diff
@@ -0,0 +1,31 @@
+name: Benchmark Test
+
+on:
+  pull_request:
+    branches:
+      - main
+
+jobs:
+  benchmark:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Install Go
+        uses: actions/setup-go@v5
+        with:
+          go-version-file: go.mod
+
+      - name: Run benchmark
+        working-directory: test/e2e
+        run: |
+            go test -run=^$ -bench=. -benchmem -count=10 > ../../new.txt
+
+      - name: Compare with baseline
+        run: |
+          go install golang.org/x/perf/cmd/benchstat@latest
+          benchstat benchmarks/baseline.txt new.txt
```

Comment:
benchstat is nice to get comparisons, but we'd have to be careful to somehow generate and commit a baseline from GH CI's system to avoid issues like "Jian's laptop is beefier than Joe's, and both are beefier than a GHA VM"

Reply:
Yeah